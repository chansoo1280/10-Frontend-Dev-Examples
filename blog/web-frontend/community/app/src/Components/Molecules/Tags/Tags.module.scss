@import "@Styles/scss/default.scss";

.tags {
    &__btn {
        &--magenta {
            @include set-color(#EB2F96, #FFF0F6, #FFADD2)
        }

        &--green {
            @include set-color(#52C41A, #F6FFED, #B7EB8F)
        }

        &--red {
            @include set-color(#F5222D, #FFF1F0, #FFA39E)
        }

        &--volcano {
            @include set-color(#FA541C, #FFF2E8, #FFBB96)
        }

        &--orange {
            @include set-color(#FA8C16, #FFF7E6, #FFD591)
        }

        &--gold {
            @include set-color(#FAAD14, #FFFBE6, #FFE58F)
        }

        &--lime {
            @include set-color(#A0D911, #FCFFE6, #EAFF8F)
        }

        &--cyan {
            @include set-color(#13C2C2, #E6FFFB, #87E8DE)
        }

        &--blue {
            @include set-color(#1890FF, #E6F7FF, #91D5FF)
        }

        &--geekblue {
            @include set-color(#2F54EB, #F0F5FF, #ADC6FF)
        }

        &--purple {
            @include set-color(#722ED1, #F9F0FF, #D3ADF7)
        }
    }
}