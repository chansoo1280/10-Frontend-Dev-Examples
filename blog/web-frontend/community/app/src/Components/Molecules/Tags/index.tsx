// #region Global Imports
import React, { useEffect, useRef, useState } from "react"
import classNames from "classnames"
// #endregion Global Imports

// #region Local Imports
import { defaultProps, Space, Input, Button, Icon } from "@Components"
import styles from "./Tags.module.scss"
// #endregion Local Imports

interface BaseTag {
    title: string
    color?: "magenta" | "green" | "red" | "volcano" | "orange" | "gold" | "lime" | "cyan" | "blue" | "geekblue" | "purple"
}
interface deletableTag extends BaseTag {
    type: "deletable"
}
interface checkableTag extends BaseTag {
    type: "checkable"
    checked: boolean
}
export type Tag = deletableTag | checkableTag
interface TagsProps extends defaultProps {
    tagList: Tag[]
    onClick: (tag: Tag) => void
    onAdd?: (text: string) => void
}

const Tags = (props: TagsProps): JSX.Element => {
    const { tagList, show, className, onClick, onAdd, ...rest } = props

    const inputRef = useRef<HTMLInputElement>(null)
    const [openAddInput, setOpenAddInput] = useState(false)
    useEffect(() => {
        if (openAddInput === true) {
            inputRef.current?.focus()
        } else {
            setAddInput("")
        }
    }, [openAddInput])
    const [addInput, setAddInput] = useState("")
    const prefixCls = "tags"
    const classes = classNames(
        styles[`${prefixCls}`],
        {
            [styles[`${prefixCls}--hide`]]: show === false,
        },
        className,
    )
    return (
        <Space className={classes} {...rest}>
            {tagList.map((tag) => (
                <Button
                    size="small"
                    type={tag.type === "checkable" && tag.checked === true ? "primary" : "secondary"}
                    onClick={() => {
                        onClick(tag)
                    }}
                    key={tag.title}
                    className={classNames(styles[`${prefixCls}__btn`], {
                        [styles[`${prefixCls}__btn--${tag.color}`]]: tag.color && ((tag.type === "checkable" && !tag.checked) || tag.type === "deletable"),
                    })}
                >
                    {tag.title}
                    <Icon show={tag.type === "deletable"} iconName="xi-close" />
                </Button>
            ))}
            {onAdd && openAddInput ? (
                <Input
                    style={{
                        width: "95px",
                    }}
                    ref={inputRef}
                    onBlur={() => {
                        setOpenAddInput(false)
                    }}
                    onEnter={() => {
                        onAdd(addInput)
                        setOpenAddInput(false)
                    }}
                    size="small"
                    value={addInput}
                    onChange={(e) => {
                        setAddInput(e.currentTarget.value)
                    }}
                />
            ) : (
                <Button
                    size="small"
                    type={"secondary"}
                    onClick={() => {
                        setOpenAddInput(true)
                    }}
                    className={classNames(styles[`${prefixCls}__btn-add`])}
                >
                    <Icon iconName="xi-plus" />
                    New Tag
                </Button>
            )}
        </Space>
    )
}
export default Tags
