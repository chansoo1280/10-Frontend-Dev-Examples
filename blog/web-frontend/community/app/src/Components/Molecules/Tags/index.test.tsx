import { fireEvent, render, screen } from "@testing-library/react"
import Tags from "./"
describe("Tags", () => {
    it("renders tags unchanged", () => {
        const handleClick = jest.fn()
        const { container } = render(<Tags onClick={handleClick} tagList={[{ title: "Tag 1", checked: false, type: "checkable" }]} />)
        expect(container).toMatchSnapshot()
    })
    it("should render a tags with the class", () => {
        const handleClick = jest.fn()
        render(<Tags data-testid="tags" onClick={handleClick} tagList={[{ title: "Tag 1", checked: false, type: "checkable" }]} />)

        const tags = screen.getByTestId("tags")
        expect(tags).toBeInTheDocument()
        expect(tags).toHaveClass("tags")
    })
    it("should render a tags, checks for onClick attribute", () => {
        const handleClick = jest.fn()
        render(
            <Tags
                data-testid="tags"
                onClick={handleClick}
                tagList={[
                    { title: "Tag 1", checked: false, type: "checkable" },
                    { title: "Tag 2", type: "deletable" },
                ]}
            />,
        )

        const button = screen.getByText("Tag 1")
        fireEvent.click(button)
        expect(handleClick).toBeCalledTimes(1)
    })
})
