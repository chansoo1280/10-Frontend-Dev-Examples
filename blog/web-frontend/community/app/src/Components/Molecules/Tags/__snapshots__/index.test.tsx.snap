// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags renders tags unchanged 1`] = `
<div>
  <div
    class="space tags"
  >
    <button
      class="btn btn--primary"
      data-click-animating="false"
      type="button"
    >
      Tag 1
    </button>
  </div>
</div>
`;
